sdks/java/extensions/ordered/src/main/java/org/apache/beam/sdk/extensions/ordered/OrderedEventProcessor.java

```diff
@@ -0,0 +1,657 @@
+    extends PTransform<
+        PCollection<KV<EventKeyT, KV<Long, EventT>>>,
+        OrderedEventProcessorResult<EventKeyT, ResultT, EventT>> {
+
+  public static <
+          EventTypeT,
+          EventKeyTypeT,
+          ResultTypeT,
+          StateTypeT extends MutableState<EventTypeT, ResultTypeT>>
+      OrderedEventProcessor<EventTypeT, EventKeyTypeT, ResultTypeT, StateTypeT> create(
+          OrderedProcessingHandler<EventTypeT, EventKeyTypeT, StateTypeT, ResultTypeT> handler) {
+    return new AutoValue_OrderedEventProcessor<>(handler);
+  }
+
+  @Nullable
+  abstract OrderedProcessingHandler<EventT, EventKeyT, StateT, ResultT> getHandler();
+
+  @Override
+  public OrderedEventProcessorResult<EventKeyT, ResultT, EventT> expand(
+      PCollection<KV<EventKeyT, KV<Long, EventT>>> input) {
+    final TupleTag<KV<EventKeyT, ResultT>> mainOutput =
+        new TupleTag<KV<EventKeyT, ResultT>>("mainOutput") {};
+    final TupleTag<KV<EventKeyT, OrderedProcessingStatus>> statusOutput =
+        new TupleTag<KV<EventKeyT, OrderedProcessingStatus>>("status") {};
+
+    final TupleTag<KV<EventKeyT, KV<Long, UnprocessedEvent<EventT>>>> unprocessedEventOutput =
+        new TupleTag<KV<EventKeyT, KV<Long, UnprocessedEvent<EventT>>>>("unprocessed-events") {};
+
+    OrderedProcessingHandler<EventT, EventKeyT, StateT, ResultT> handler = getHandler();
+    Pipeline pipeline = input.getPipeline();
+
+    Coder<EventKeyT> keyCoder;
+    try {
+      keyCoder = handler.getKeyCoder(pipeline, input.getCoder());
+    } catch (CannotProvideCoderException e) {
+      throw new RuntimeException("Unable to get key coder", e);
+    }
+
+    Coder<EventT> eventCoder;
+    try {
+      eventCoder = handler.getEventCoder(pipeline, input.getCoder());
+    } catch (CannotProvideCoderException e) {
+      throw new RuntimeException("Unable to get event coder", e);
+    }
+
+    Coder<StateT> stateCoder;
+    try {
+      stateCoder = handler.getStateCoder(pipeline);
+    } catch (CannotProvideCoderException e) {
+      throw new RuntimeException("Unable to get state coder", e);
+    }
+
+    Coder<ResultT> resultCoder;
+    try {
+      resultCoder = handler.getResultCoder(pipeline);
+    } catch (CannotProvideCoderException e) {
+      throw new RuntimeException("Unable to get result coder", e);
+    }
+
+    PCollectionTuple processingResult =
+        input.apply(
+            ParDo.of(
+                    new OrderedProcessorDoFn<>(
+                        handler.getEventExaminer(),
+                        eventCoder,
+                        stateCoder,
+                        keyCoder,
+                        mainOutput,
+                        statusOutput,
+                        handler.getStatusUpdateFrequency(),
+                        unprocessedEventOutput,
+                        handler.isProduceStatusUpdateOnEveryEvent(),
+                        input.isBounded() == IsBounded.BOUNDED
```

Comment:
Why do we ignore `getMaxOutputElementsPerBundle` for bounded inputs?

Reply:
Removed that conditional. I will add unit tests for batch pipelines to make sure things are still working correctly with timers.

Reply:
Added batch tests.